File: jas_ocaml/lib/panels/color_panel_view.ml
(** Color panel body: swatches, fill/stroke widget, sliders, hex input, color bar. *)

open Workspace_layout

let apply_dark_css w css_str =
  let css = new GObj.css_provider (GtkData.CssProvider.create ()) in
  css#load_from_data css_str;
  w#misc#style_context#add_provider css 600

(** Panel-local mutable color state. *)
type panel_state = {
  mutable h : float;   (* 0..360 *)
  mutable s : float;   (* 0..100 *)
  mutable b : float;   (* 0..100 *)
  mutable r : float;   (* 0..255 *)
  mutable g : float;   (* 0..255 *)
  mutable bl : float;  (* 0..255 *)
  mutable c : float;   (* 0..100 *)
  mutable m : float;   (* 0..100 *)
  mutable y : float;   (* 0..100 *)
  mutable k : float;   (* 0..100 *)
  mutable hex : string;
}

let make_panel_state () =
  { h = 0.0; s = 0.0; b = 100.0;
    r = 255.0; g = 255.0; bl = 255.0;
    c = 0.0; m = 0.0; y = 0.0; k = 0.0;
    hex = "ffffff" }

let sync_from_color ps (color : Element.color) =
  let (rv, gv, bv, _) = Element.color_to_rgba color in
  ps.r <- Float.round (rv *. 255.0);
  ps.g <- Float.round (gv *. 255.0);
  ps.bl <- Float.round (bv *. 255.0);
  let (hv, sv, brv, _) = Element.color_to_hsba color in
  ps.h <- Float.round hv;
  ps.s <- Float.round (sv *. 100.0);
  ps.b <- Float.round (brv *. 100.0);
  let (cv, mv, yv, kv, _) = Element.color_to_cmyka color in
  ps.c <- Float.round (cv *. 100.0);
  ps.m <- Float.round (mv *. 100.0);
  ps.y <- Float.round (yv *. 100.0);
  ps.k <- Float.round (kv *. 100.0);
  ps.hex <- Element.color_to_hex color

let ps_to_color ps mode =
  match mode with
  | Hsb_mode -> Element.color_hsb ps.h (ps.s /. 100.0) (ps.b /. 100.0)
  | Rgb_mode | Web_safe_rgb -> Element.color_rgb (ps.r /. 255.0) (ps.g /. 255.0) (ps.bl /. 255.0)
  | Cmyk_mode -> Element.color_cmyk (ps.c /. 100.0) (ps.m /. 100.0) (ps.y /. 100.0) (ps.k /. 100.0)
  | Grayscale ->
    let v = 1.0 -. ps.k /. 100.0 in
    Element.color_rgb v v v

let get_field ps field =
  match field with
  | "h" -> ps.h | "s" -> ps.s | "b" -> ps.b
  | "r" -> ps.r | "g" -> ps.g | "bl" -> ps.bl
  | "c" -> ps.c | "m" -> ps.m | "y" -> ps.y | "k" -> ps.k
  | _ -> 0.0

let set_field ps field v =
  match field with
  | "h" -> ps.h <- v | "s" -> ps.s <- v | "b" -> ps.b <- v
  | "r" -> ps.r <- v | "g" -> ps.g <- v | "bl" -> ps.bl <- v
  | "c" -> ps.c <- v | "m" -> ps.m <- v | "y" -> ps.y <- v | "k" -> ps.k <- v
  | _ -> ()

(** Build the color panel body widgets into the given container.
    Returns a refresh function to call when external state changes. *)
let create
    ~(packing : GObj.widget -> unit)
    ~(layout : workspace_layout)
    ~(get_model : unit -> Model.model)
    ~(get_fill_on_top : unit -> bool)
    ~(rebuild : unit -> unit)
    ~theme_text ~theme_text_dim ~theme_bg_dark ~theme_border
    () =
  let ps = make_panel_state () in
  let last_synced_hex = ref "" in

  let vbox = GPack.vbox ~spacing:4 ~packing () in
  apply_dark_css vbox (Printf.sprintf "box { padding: 4px; }");

  let mode () = layout.color_panel_mode in
  let fill_on_top () = get_fill_on_top () in
  let active_color () =
    let m = get_model () in
    if fill_on_top () then
      Option.map (fun (f : Element.fill) -> f.fill_color) m#default_fill
    else
      Option.map (fun (s : Element.stroke) -> s.stroke_color) m#default_stroke
  in

  (* Sync panel state from active color *)
  let sync () =
    let hex = match active_color () with Some c -> Element.color_to_hex c | None -> "" in
    if hex <> !last_synced_hex then begin
      (match active_color () with Some c -> sync_from_color ps c | None -> ());
      last_synced_hex := hex
    end
  in
  sync ();

  (* ── Row 1: Swatches ── *)
  let swatches_box = GPack.hbox ~spacing:2 ~packing:(vbox#pack ~expand:false) () in

  (* None shortcut *)
  let none_btn = GButton.button ~label:"\xE2\x88\x85" ~packing:(swatches_box#pack ~expand:false) () in
  apply_dark_css none_btn (Printf.sprintf "button { font-size: 12px; color: red; background: %s; border: 1px solid %s; min-width: 16px; min-height: 16px; padding: 0; }" !theme_bg_dark !theme_border);
  none_btn#connect#clicked ~callback:(fun () ->
    let m = get_model () in
    if fill_on_top () then begin
      m#set_default_fill None;
      if not (Document.PathMap.is_empty m#document.Document.selection) then begin
        m#snapshot;
        let ctrl = Controller.create ~model:m () in
        ctrl#set_selection_fill None
      end
    end else begin
      m#set_default_stroke None;
      if not (Document.PathMap.is_empty m#document.Document.selection) then begin
        m#snapshot;
        let ctrl = Controller.create ~model:m () in
        ctrl#set_selection_stroke None
      end
    end;
    rebuild ()
  ) |> ignore;

  (* Black shortcut *)
  let black_btn = GButton.button ~packing:(swatches_box#pack ~expand:false) () in
  apply_dark_css black_btn "button { background: #000000; min-width: 16px; min-height: 16px; padding: 0; border: 1px solid #888; }";
  black_btn#connect#clicked ~callback:(fun () ->
    Panel_menu.set_active_color Element.black ~fill_on_top:(fill_on_top ()) (get_model ());
    rebuild ()
  ) |> ignore;

  (* White shortcut *)
  let white_btn = GButton.button ~packing:(swatches_box#pack ~expand:false) () in
  apply_dark_css white_btn "button { background: #ffffff; min-width: 16px; min-height: 16px; padding: 0; border: 1px solid #888; }";
  white_btn#connect#clicked ~callback:(fun () ->
    Panel_menu.set_active_color Element.white ~fill_on_top:(fill_on_top ()) (get_model ());
    rebuild ()
  ) |> ignore;

  (* Separator *)
  let _sep = GMisc.separator `VERTICAL ~packing:(swatches_box#pack ~expand:false) () in

  (* Recent color slots *)
  let recent_btns = Array.init 10 (fun _i ->
    let btn = GButton.button ~packing:(swatches_box#pack ~expand:false) () in
    apply_dark_css btn (Printf.sprintf "button { background: transparent; min-width: 16px; min-height: 16px; padding: 0; border: 1px solid %s; }" !theme_border);
    btn
  ) in

  let update_recent_swatches () =
    let m = get_model () in
    let rc = m#recent_colors in
    Array.iteri (fun i btn ->
      match List.nth_opt rc i with
      | Some hex ->
        apply_dark_css btn (Printf.sprintf "button { background: #%s; min-width: 16px; min-height: 16px; padding: 0; border: 1px solid #888; }" hex)
      | None ->
        apply_dark_css btn (Printf.sprintf "button { background: transparent; min-width: 16px; min-height: 16px; padding: 0; border: 1px solid %s; }" !theme_border)
    ) recent_btns
  in
  update_recent_swatches ();

  Array.iteri (fun i btn ->
    btn#connect#clicked ~callback:(fun () ->
      let m = get_model () in
      match List.nth_opt m#recent_colors i with
      | Some hex ->
        (match Element.color_from_hex hex with
         | Some c ->
           Panel_menu.set_active_color c ~fill_on_top:(fill_on_top ()) m;
           rebuild ()
         | None -> ())
      | None -> ()
    ) |> ignore
  ) recent_btns;

  (* ── Row 2: Sliders ── *)
  let sliders_box = GPack.vbox ~spacing:2 ~packing:(vbox#pack ~expand:false) () in

  let slider_rows = ref [] in

  let add_slider_row label field min_val max_val step suffix =
    let row = GPack.hbox ~spacing:4 () in
    let _lbl = GMisc.label ~text:label ~packing:(row#pack ~expand:false) () in
    apply_dark_css _lbl (Printf.sprintf "label { color: %s; font-size: 10px; min-width: 10px; }" !theme_text);
    let adj = GData.adjustment ~value:(get_field ps field) ~lower:min_val ~upper:max_val ~step_incr:step ~page_incr:(step *. 10.0) () in
    let scale = GRange.scale `HORIZONTAL ~adjustment:adj ~draw_value:false ~packing:(row#pack ~expand:true ~fill:true) () in
    scale#misc#set_size_request ~height:18 ();
    apply_dark_css scale (Printf.sprintf "scale { min-height: 14px; }");
    let val_label = GMisc.label ~text:(string_of_int (int_of_float (get_field ps field)))
      ~packing:(row#pack ~expand:false) () in
    apply_dark_css val_label (Printf.sprintf "label { color: %s; font-size: 10px; min-width: 30px; }" !theme_text);
    (match suffix with
     | Some sfx ->
       let _sfx_lbl = GMisc.label ~text:sfx ~packing:(row#pack ~expand:false) () in
       apply_dark_css _sfx_lbl (Printf.sprintf "label { color: %s; font-size: 10px; }" !theme_text_dim)
     | None -> ());
    adj#connect#value_changed ~callback:(fun () ->
      let v = adj#value in
      set_field ps field v;
      let color = ps_to_color ps (mode ()) in
      sync_from_color ps color;
      let m2 = mode () in
      set_field ps field v;
      ignore m2;
      last_synced_hex := Element.color_to_hex color;
      val_label#set_text (string_of_int (int_of_float v));
      Panel_menu.set_active_color_live color ~fill_on_top:(fill_on_top ()) (get_model ())
    ) |> ignore;
    slider_rows := (row, field, adj, val_label) :: !slider_rows;
    row
  in

  let rebuild_sliders () =
    List.iter (fun w -> w#destroy ()) sliders_box#children;
    slider_rows := [];
    (match mode () with
     | Grayscale ->
       sliders_box#pack ~expand:false (add_slider_row "K" "k" 0.0 100.0 1.0 (Some "%"))#coerce
     | Hsb_mode ->
       sliders_box#pack ~expand:false (add_slider_row "H" "h" 0.0 360.0 1.0 (Some "\xC2\xB0"))#coerce;
       sliders_box#pack ~expand:false (add_slider_row "S" "s" 0.0 100.0 1.0 (Some "%"))#coerce;
       sliders_box#pack ~expand:false (add_slider_row "B" "b" 0.0 100.0 1.0 (Some "%"))#coerce
     | Rgb_mode ->
       sliders_box#pack ~expand:false (add_slider_row "R" "r" 0.0 255.0 1.0 None)#coerce;
       sliders_box#pack ~expand:false (add_slider_row "G" "g" 0.0 255.0 1.0 None)#coerce;
       sliders_box#pack ~expand:false (add_slider_row "B" "bl" 0.0 255.0 1.0 None)#coerce
     | Cmyk_mode ->
       sliders_box#pack ~expand:false (add_slider_row "C" "c" 0.0 100.0 1.0 (Some "%"))#coerce;
       sliders_box#pack ~expand:false (add_slider_row "M" "m" 0.0 100.0 1.0 (Some "%"))#coerce;
       sliders_box#pack ~expand:false (add_slider_row "Y" "y" 0.0 100.0 1.0 (Some "%"))#coerce;
       sliders_box#pack ~expand:false (add_slider_row "K" "k" 0.0 100.0 1.0 (Some "%"))#coerce
     | Web_safe_rgb ->
       sliders_box#pack ~expand:false (add_slider_row "R" "r" 0.0 255.0 51.0 None)#coerce;
       sliders_box#pack ~expand:false (add_slider_row "G" "g" 0.0 255.0 51.0 None)#coerce;
       sliders_box#pack ~expand:false (add_slider_row "B" "bl" 0.0 255.0 51.0 None)#coerce);
    sliders_box#misc#show_all ()
  in
  rebuild_sliders ();

  (* ── Row 3: Hex input ── *)
  let hex_box = GPack.hbox ~spacing:2 ~packing:(vbox#pack ~expand:false) () in
  let _hash = GMisc.label ~text:"#" ~packing:(hex_box#pack ~expand:false) () in
  apply_dark_css _hash (Printf.sprintf "label { color: %s; font-size: 10px; }" !theme_text);
  let hex_entry = GEdit.entry ~text:ps.hex ~width_chars:8 ~packing:(hex_box#pack ~expand:false) () in
  apply_dark_css hex_entry (Printf.sprintf "entry { font-size: 10px; font-family: monospace; background: %s; color: %s; border: 1px solid %s; }" !theme_bg_dark !theme_text !theme_border);
  hex_entry#connect#activate ~callback:(fun () ->
    let raw = String.trim hex_entry#text in
    let raw = if String.length raw > 0 && raw.[0] = '#' then String.sub raw 1 (String.length raw - 1) else raw in
    if String.length raw = 6 then
      match Element.color_from_hex raw with
      | Some color ->
        sync_from_color ps color;
        last_synced_hex := Element.color_to_hex color;
        Panel_menu.set_active_color color ~fill_on_top:(fill_on_top ()) (get_model ());
        rebuild ()
      | None -> ()
  ) |> ignore;

  (* ── Color bar ── *)
  let bar_area = GMisc.drawing_area ~packing:(vbox#pack ~expand:false) () in
  bar_area#misc#set_size_request ~height:64 ();
  bar_area#event#add [`BUTTON_PRESS; `BUTTON_RELEASE; `BUTTON_MOTION];

  bar_area#misc#connect#draw ~callback:(fun cr ->
    let alloc = bar_area#misc#allocation in
    let w = alloc.Gtk.width in
    let h = alloc.Gtk.height in
    if w > 0 && h > 0 then begin
      let surface = Cairo.Image.create Cairo.Image.RGB24 ~w ~h in
      let data = Cairo.Image.get_data32 surface in
      let mid_y = float_of_int h /. 2.0 in
      for y = 0 to h - 1 do
        let yf = float_of_int y in
        let (sat, br) = if yf <= mid_y then
          let t = yf /. mid_y in
          (t, 1.0 -. t *. 0.2)
        else
          let t = (yf -. mid_y) /. (float_of_int h -. mid_y) in
          (1.0, 0.8 *. (1.0 -. t))
        in
        for x = 0 to w - 1 do
          let hue = 360.0 *. float_of_int x /. float_of_int w in
          let c = Element.color_hsb hue sat br in
          let (rv, gv, bv, _) = Element.color_to_rgba c in
          let ri = int_of_float (rv *. 255.0) in
          let gi = int_of_float (gv *. 255.0) in
          let bi = int_of_float (bv *. 255.0) in
          data.{y, x} <- Int32.of_int (ri lsl 16 lor gi lsl 8 lor bi)
        done
      done;
      Cairo.set_source_surface cr surface ~x:0.0 ~y:0.0;
      Cairo.paint cr
    end;
    true
  ) |> ignore;

  let apply_bar_point ev commit =
    let alloc = bar_area#misc#allocation in
    let w = float_of_int alloc.Gtk.width in
    let h = float_of_int alloc.Gtk.height in
    let x = Float.max 0.0 (Float.min (GdkEvent.Button.x ev) (w -. 1.0)) in
    let y = Float.max 0.0 (Float.min (GdkEvent.Button.y ev) (h -. 1.0)) in
    let mid_y = h /. 2.0 in
    let hue = 360.0 *. x /. w in
    let (sat, br) = if y <= mid_y then
      let t = y /. mid_y in
      (t *. 100.0, 100.0 -. t *. 20.0)
    else
      let t = (y -. mid_y) /. (h -. mid_y) in
      (100.0, 80.0 *. (1.0 -. t))
    in
    let color = Element.color_hsb hue (sat /. 100.0) (br /. 100.0) in
    sync_from_color ps color;
    ps.h <- Float.round hue;
    ps.s <- Float.round sat;
    ps.b <- Float.round br;
    last_synced_hex := Element.color_to_hex color;
    if commit then
      Panel_menu.set_active_color color ~fill_on_top:(fill_on_top ()) (get_model ())
    else
      Panel_menu.set_active_color_live color ~fill_on_top:(fill_on_top ()) (get_model ());
    rebuild ()
  in

  bar_area#event#connect#button_press ~callback:(fun ev ->
    if GdkEvent.Button.button ev = 1 then begin
      apply_bar_point ev false;
      true
    end else false
  ) |> ignore;

  bar_area#event#connect#button_release ~callback:(fun ev ->
    if GdkEvent.Button.button ev = 1 then begin
      apply_bar_point ev true;
      true
    end else false
  ) |> ignore;

  (* Return a refresh function *)
  let refresh () =
    sync ();
    hex_entry#set_text ps.hex;
    update_recent_swatches ();
    rebuild_sliders ();
    bar_area#misc#queue_draw ()
  in
  refresh

File: jas/panels/color_panel_view.py
"""Color panel body widget: swatches, sliders, hex input, color bar."""

from __future__ import annotations

from PySide6.QtWidgets import (QWidget, QVBoxLayout, QHBoxLayout, QLabel,
                                QSlider, QLineEdit, QPushButton)
from PySide6.QtCore import Qt
from PySide6.QtGui import QColor, QPainter, QImage, QMouseEvent

from geometry.element import Color, Fill, Stroke
from panels.panel_menu import (set_active_color, set_active_color_live,
                                push_recent_color, COLOR_MODE_COMMANDS)
from workspace.dock_panel import (THEME_TEXT, THEME_TEXT_DIM, THEME_TEXT_BODY,
                                   THEME_BG_DARK, THEME_BORDER)


# ---------------------------------------------------------------------------
# Panel color state
# ---------------------------------------------------------------------------

class PanelColorState:
    def __init__(self):
        self.h = 0.0; self.s = 0.0; self.b = 100.0
        self.r = 255.0; self.g = 255.0; self.bl = 255.0
        self.c = 0.0; self.m = 0.0; self.y = 0.0; self.k = 0.0
        self.hex = "ffffff"

    def sync_from_color(self, color: Color):
        r, g, b, _ = color.to_rgba()
        self.r = round(r * 255); self.g = round(g * 255); self.bl = round(b * 255)
        h, s, br, _ = color.to_hsba()
        self.h = round(h); self.s = round(s * 100); self.b = round(br * 100)
        c, m, y, k, _ = color.to_cmyka()
        self.c = round(c * 100); self.m = round(m * 100)
        self.y = round(y * 100); self.k = round(k * 100)
        self.hex = color.to_hex()

    def to_color(self, mode: str) -> Color:
        if mode == "hsb":
            return Color.hsb(self.h, self.s / 100, self.b / 100)
        elif mode in ("rgb", "web_safe_rgb"):
            return Color.rgb(self.r / 255, self.g / 255, self.bl / 255)
        elif mode == "cmyk":
            return Color.cmyk(self.c / 100, self.m / 100, self.y / 100, self.k / 100)
        else:  # grayscale
            v = 1.0 - self.k / 100
            return Color.rgb(v, v, v)

    def get(self, field: str) -> float:
        return getattr(self, field, 0.0)

    def set(self, field: str, val: float):
        setattr(self, field, val)


# ---------------------------------------------------------------------------
# Color bar widget
# ---------------------------------------------------------------------------

class ColorBarWidget(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setFixedHeight(64)
        self.setMinimumWidth(50)
        self._on_click = None
        self._on_drag = None
        self._on_release = None
        self._dragging = False

    def set_callbacks(self, on_click, on_drag, on_release):
        self._on_click = on_click
        self._on_drag = on_drag
        self._on_release = on_release

    def paintEvent(self, event):
        painter = QPainter(self)
        w, h = self.width(), self.height()
        if w <= 0 or h <= 0:
            return
        img = QImage(w, h, QImage.Format_RGB32)
        mid_y = h / 2.0
        for y in range(h):
            yf = float(y)
            if yf <= mid_y:
                t = yf / mid_y
                sat, br = t, 1.0 - t * 0.2
            else:
                t = (yf - mid_y) / (h - mid_y) if h > mid_y else 0
                sat, br = 1.0, 0.8 * (1.0 - t)
            for x in range(w):
                hue = 360.0 * x / w
                c = Color.hsb(hue, sat, br)
                rv, gv, bv, _ = c.to_rgba()
                img.setPixelColor(x, y, QColor(int(rv * 255), int(gv * 255), int(bv * 255)))
        painter.drawImage(0, 0, img)

    def mousePressEvent(self, event: QMouseEvent):
        if event.button() == Qt.LeftButton:
            self._dragging = True
            if self._on_click:
                self._on_click(event.position().x(), event.position().y())

    def mouseMoveEvent(self, event: QMouseEvent):
        if self._dragging and self._on_drag:
            self._on_drag(event.position().x(), event.position().y())

    def mouseReleaseEvent(self, event: QMouseEvent):
        if self._dragging:
            self._dragging = False
            if self._on_release:
                self._on_release(event.position().x(), event.position().y())


# ---------------------------------------------------------------------------
# Color panel view
# ---------------------------------------------------------------------------

class ColorPanelView(QWidget):
    def __init__(self, layout, get_model, rebuild_fn, parent=None):
        super().__init__(parent)
        self._layout = layout
        self._get_model = get_model
        self._rebuild_fn = rebuild_fn
        self._ps = PanelColorState()
        self._last_synced_hex = ""
        self._slider_widgets = []

        vbox = QVBoxLayout(self)
        vbox.setContentsMargins(4, 4, 4, 4)
        vbox.setSpacing(6)

        self._build_swatches(vbox)
        self._sliders_container = QWidget()
        self._sliders_layout = QVBoxLayout(self._sliders_container)
        self._sliders_layout.setContentsMargins(0, 0, 0, 0)
        self._sliders_layout.setSpacing(2)
        vbox.addWidget(self._sliders_container)
        self._build_sliders()
        self._build_hex_row(vbox)
        self._build_color_bar(vbox)

        self._sync()

    @property
    def _mode(self): return self._layout.color_panel_mode

    @property
    def _model(self): return self._get_model()

    @property
    def _fill_on_top(self):
        m = self._model
        return m.fill_on_top if m else True

    @property
    def _active_color(self):
        m = self._model
        if m is None:
            return None
        if m.fill_on_top:
            return m.default_fill.color if m.default_fill else None
        else:
            return m.default_stroke.color if m.default_stroke else None

    def _sync(self):
        color = self._active_color
        hex_val = color.to_hex() if color else ""
        if hex_val != self._last_synced_hex:
            if color:
                self._ps.sync_from_color(color)
            self._last_synced_hex = hex_val
            self._update_ui()

    def _update_ui(self):
        if hasattr(self, '_hex_entry'):
            self._hex_entry.setText(self._ps.hex)
        self._update_recent_swatches()

    # -- Swatches row --

    def _build_swatches(self, parent_layout):
        row = QHBoxLayout()
        row.setSpacing(2)

        # None
        none_btn = QPushButton("\u2205")
        none_btn.setFixedSize(16, 16)
        none_btn.setStyleSheet(f"font-size:12px; color:red; background:{THEME_BG_DARK}; border:1px solid {THEME_BORDER}; padding:0;")
        none_btn.clicked.connect(self._on_none)
        row.addWidget(none_btn)

        # Black
        black_btn = QPushButton()
        black_btn.setFixedSize(16, 16)
        black_btn.setStyleSheet("background:#000; border:1px solid #888; padding:0;")
        black_btn.clicked.connect(lambda: self._set_color(Color.BLACK))
        row.addWidget(black_btn)

        # White
        white_btn = QPushButton()
        white_btn.setFixedSize(16, 16)
        white_btn.setStyleSheet("background:#fff; border:1px solid #888; padding:0;")
        white_btn.clicked.connect(lambda: self._set_color(Color.WHITE))
        row.addWidget(white_btn)

        # Separator
        sep = QWidget()
        sep.setFixedSize(1, 16)
        sep.setStyleSheet(f"background:{THEME_BORDER};")
        row.addWidget(sep)

        # Recent swatches
        self._recent_btns = []
        for i in range(10):
            btn = QPushButton()
            btn.setFixedSize(16, 16)
            btn.setStyleSheet(f"background:transparent; border:1px solid {THEME_BORDER}; padding:0;")
            idx = i
            btn.clicked.connect(lambda checked=False, ii=idx: self._on_recent(ii))
            row.addWidget(btn)
            self._recent_btns.append(btn)

        row.addStretch()
        parent_layout.addLayout(row)

    def _update_recent_swatches(self):
        rc = self._model.recent_colors
        for i, btn in enumerate(self._recent_btns):
            if i < len(rc):
                btn.setStyleSheet(f"background:#{rc[i]}; border:1px solid #888; padding:0;")
            else:
                btn.setStyleSheet(f"background:transparent; border:1px solid {THEME_BORDER}; padding:0;")

    def _on_none(self):
        m = self._model
        if m is None:
            return
        if m.fill_on_top:
            m.default_fill = None
        else:
            m.default_stroke = None
        self._rebuild_fn()

    def _on_recent(self, idx):
        rc = self._model.recent_colors
        if idx < len(rc):
            color = Color.from_hex(rc[idx])
            if color:
                self._set_color(color)

    def _set_color(self, color):
        m = self._model
        if m is None:
            return
        set_active_color(color, m)
        self._rebuild_fn()

    # -- Sliders --

    def _build_sliders(self):
        # Clear
        while self._sliders_layout.count():
            item = self._sliders_layout.takeAt(0)
            if item.widget():
                item.widget().deleteLater()
        self._slider_widgets = []

        mode = self._mode
        if mode == "grayscale":
            self._add_slider("K", "k", 0, 100, 1, "%")
        elif mode == "hsb":
            self._add_slider("H", "h", 0, 360, 1, "\u00B0")
            self._add_slider("S", "s", 0, 100, 1, "%")
            self._add_slider("B", "b", 0, 100, 1, "%")
        elif mode == "rgb":
            self._add_slider("R", "r", 0, 255, 1, "")
            self._add_slider("G", "g", 0, 255, 1, "")
            self._add_slider("B", "bl", 0, 255, 1, "")
        elif mode == "cmyk":
            self._add_slider("C", "c", 0, 100, 1, "%")
            self._add_slider("M", "m", 0, 100, 1, "%")
            self._add_slider("Y", "y", 0, 100, 1, "%")
            self._add_slider("K", "k", 0, 100, 1, "%")
        elif mode == "web_safe_rgb":
            self._add_slider("R", "r", 0, 255, 51, "")
            self._add_slider("G", "g", 0, 255, 51, "")
            self._add_slider("B", "bl", 0, 255, 51, "")

    def _add_slider(self, label, field, min_val, max_val, step, suffix):
        row = QWidget()
        hbox = QHBoxLayout(row)
        hbox.setContentsMargins(0, 0, 0, 0)
        hbox.setSpacing(4)

        lbl = QLabel(label)
        lbl.setFixedWidth(10)
        lbl.setStyleSheet(f"color:{THEME_TEXT}; font-size:10px;")
        lbl.setAlignment(Qt.AlignRight | Qt.AlignVCenter)
        hbox.addWidget(lbl)

        slider = QSlider(Qt.Horizontal)
        slider.setMinimum(min_val)
        slider.setMaximum(max_val)
        slider.setSingleStep(step)
        slider.setValue(int(self._ps.get(field)))
        hbox.addWidget(slider, 1)

        val_lbl = QLabel(str(int(self._ps.get(field))))
        val_lbl.setFixedWidth(30)
        val_lbl.setStyleSheet(f"color:{THEME_TEXT}; font-size:10px;")
        val_lbl.setAlignment(Qt.AlignRight | Qt.AlignVCenter)
        hbox.addWidget(val_lbl)

        if suffix:
            sfx_lbl = QLabel(suffix)
            sfx_lbl.setStyleSheet(f"color:{THEME_TEXT_DIM}; font-size:10px;")
            hbox.addWidget(sfx_lbl)

        slider.valueChanged.connect(
            lambda v, f=field, vl=val_lbl: self._on_slider(f, v, vl))

        self._sliders_layout.addWidget(row)
        self._slider_widgets.append((field, slider, val_lbl))

    def _on_slider(self, field, value, val_label):
        self._ps.set(field, float(value))
        color = self._ps.to_color(self._mode)
        self._ps.sync_from_color(color)
        self._ps.set(field, float(value))
        self._last_synced_hex = color.to_hex()
        val_label.setText(str(value))
        set_active_color_live(color, self._model)

    # -- Hex row --

    def _build_hex_row(self, parent_layout):
        row = QHBoxLayout()
        row.setSpacing(2)
        hash_lbl = QLabel("#")
        hash_lbl.setStyleSheet(f"color:{THEME_TEXT}; font-size:10px;")
        row.addWidget(hash_lbl)

        self._hex_entry = QLineEdit(self._ps.hex)
        self._hex_entry.setMaxLength(6)
        self._hex_entry.setFixedWidth(52)
        self._hex_entry.setStyleSheet(
            f"background:{THEME_BG_DARK}; color:{THEME_TEXT}; "
            f"border:1px solid {THEME_BORDER}; font-size:10px; font-family:monospace; padding:2px 4px;")
        self._hex_entry.returnPressed.connect(self._on_hex_commit)
        row.addWidget(self._hex_entry)
        row.addStretch()
        parent_layout.addLayout(row)

    def _on_hex_commit(self):
        raw = self._hex_entry.text().strip().lstrip("#")
        if len(raw) == 6 and all(c in "0123456789abcdefABCDEF" for c in raw):
            color = Color.from_hex(raw)
            if color:
                self._ps.sync_from_color(color)
                self._last_synced_hex = color.to_hex()
                set_active_color(color, self._model)
                self._rebuild_fn()

    # -- Color bar --

    def _build_color_bar(self, parent_layout):
        self._color_bar = ColorBarWidget()
        self._color_bar.set_callbacks(
            on_click=lambda x, y: self._bar_point(x, y, False),
            on_drag=lambda x, y: self._bar_point(x, y, False),
            on_release=lambda x, y: self._bar_point(x, y, True),
        )
        parent_layout.addWidget(self._color_bar)

    def _bar_point(self, x, y, commit):
        w = max(self._color_bar.width(), 1)
        h = max(self._color_bar.height(), 1)
        x = max(0.0, min(x, w - 1))
        y = max(0.0, min(y, h - 1))
        mid_y = h / 2.0
        hue = 360.0 * x / w
        if y <= mid_y:
            t = y / mid_y
            sat, br = t * 100, 100 - t * 20
        else:
            t = (y - mid_y) / (h - mid_y) if h > mid_y else 0
            sat, br = 100, 80 * (1 - t)

        color = Color.hsb(hue, sat / 100, br / 100)
        self._ps.sync_from_color(color)
        self._ps.h = round(hue)
        self._ps.s = round(sat)
        self._ps.b = round(br)
        self._last_synced_hex = color.to_hex()

        if commit:
            set_active_color(color, self._model)
            self._rebuild_fn()
        else:
            set_active_color_live(color, self._model)
